File: android/app/src/main/kotlin/com/thecityandthebike/ui/viewmodel/UserViewModel.kt
package com.thecityandthebike.ui.viewmodel

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.thecityandthebike.data.local.TokenManager
import com.thecityandthebike.data.model.ApiResult
import com.thecityandthebike.data.model.dto.SubmissionResponse
import com.thecityandthebike.data.model.dto.UserDetailResponse
import com.thecityandthebike.data.repository.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class UserState(
    val isLoading: Boolean = false,
    val userDetail: UserDetailResponse? = null,
    val submissions: List<SubmissionResponse> = emptyList(),
    val error: String? = null,
    val isLoadingMore: Boolean = false,
    val hasMorePages: Boolean = true,
    val nextCursor: String? = null,
    val isBanning: Boolean = false,
    val currentUserIsAdmin: Boolean = false
)

@HiltViewModel
class UserViewModel @Inject constructor(
    private val userRepository: UserRepository,
    private val tokenManager: TokenManager,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    val userId: String = checkNotNull(savedStateHandle["userId"])

    private val _state = MutableStateFlow(UserState(currentUserIsAdmin = tokenManager.isAdmin()))
    val state: StateFlow<UserState> = _state.asStateFlow()

    init {
        loadUser()
    }

    private fun loadUser() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, error = null)

            when (val detailResult = userRepository.getUserDetail(userId)) {
                is ApiResult.Success -> {
                    _state.value = _state.value.copy(userDetail = detailResult.data)
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isLoading = false,
                        error = detailResult.error.displayMessageOrNull
                    )
                    return@launch
                }
            }

            when (val subsResult = userRepository.getUserSubmissions(userId)) {
                is ApiResult.Success -> {
                    _state.value = _state.value.copy(
                        isLoading = false,
                        submissions = subsResult.data.items,
                        hasMorePages = subsResult.data.hasMore,
                        nextCursor = subsResult.data.nextCursor
                    )
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isLoading = false,
                        error = subsResult.error.displayMessageOrNull
                    )
                }
            }
        }
    }

    fun loadMoreSubmissions() {
        val currentState = _state.value
        if (currentState.isLoadingMore || !currentState.hasMorePages) return

        viewModelScope.launch {
            _state.value = _state.value.copy(isLoadingMore = true)

            when (val result = userRepository.getUserSubmissions(userId, cursor = _state.value.nextCursor)) {
                is ApiResult.Success -> {
                    val newSubmissions = _state.value.submissions + result.data.items
                    _state.value = _state.value.copy(
                        isLoadingMore = false,
                        submissions = newSubmissions,
                        hasMorePages = result.data.hasMore,
                        nextCursor = result.data.nextCursor
                    )
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isLoadingMore = false,
                        error = result.error.displayMessageOrNull
                    )
                }
            }
        }
    }

    fun removeSubmission(submissionId: String) {
        _state.value = _state.value.copy(
            submissions = _state.value.submissions.filter { it.submissionId != submissionId }
        )
    }

    fun banUser() = toggleBan(ban = true)

    fun unbanUser() = toggleBan(ban = false)

    private fun toggleBan(ban: Boolean) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isBanning = true)
            val result = if (ban) userRepository.banUser(userId) else userRepository.unbanUser(userId)
            when (result) {
                is ApiResult.Success -> {
                    val detail = _state.value.userDetail
                    _state.value = _state.value.copy(
                        isBanning = false,
                        userDetail = detail?.copy(isBanned = ban)
                    )
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isBanning = false,
                        error = result.error.displayMessageOrNull
                    )
                }
            }
        }
    }

    fun clearError() {
        _state.value = _state.value.copy(error = null)
    }
}

File: ios/Packages/TCATBProfile/Sources/TCATBProfile/UserViewModel.swift
import Foundation
import Observation
import TCATBModels

@Observable
public final class UserViewModel {
    public private(set) var userDetail: UserDetailResponse?
    public private(set) var submissions: [SubmissionResponse] = []
    public private(set) var isLoading = false
    public private(set) var isLoadingMore = false
    public private(set) var error: String?
    public private(set) var hasMorePages = true
    public private(set) var isBanning = false
    public let currentUserIsAdmin: Bool

    private var nextCursor: String?
    private let userId: String
    private let apiClient: ProfileAPIClient

    public init(userId: String, apiClient: ProfileAPIClient, currentUserIsAdmin: Bool = false) {
        self.userId = userId
        self.apiClient = apiClient
        self.currentUserIsAdmin = currentUserIsAdmin
    }

    public func loadUser() async {
        isLoading = true
        error = nil

        do {
            let detail = try await apiClient.getUserDetail(userId: userId)
            userDetail = detail
        } catch {
            isLoading = false
            self.error = error.localizedDescription
            return
        }

        do {
            let result = try await apiClient.getUserSubmissions(userId: userId, cursor: nil)
            submissions = result.items
            hasMorePages = result.hasMore
            nextCursor = result.nextCursor
        } catch {
            self.error = error.localizedDescription
        }

        isLoading = false
    }

    public func loadMoreSubmissions() async {
        guard !isLoadingMore, hasMorePages else { return }
        isLoadingMore = true

        do {
            let result = try await apiClient.getUserSubmissions(userId: userId, cursor: nextCursor)
            submissions += result.items
            hasMorePages = result.hasMore
            nextCursor = result.nextCursor
        } catch {
            self.error = error.localizedDescription
        }

        isLoadingMore = false
    }

    public func removeSubmission(submissionId: String) {
        submissions.removeAll { $0.submissionId == submissionId }
    }

    public func banUser() async {
        isBanning = true
        do {
            try await apiClient.banUser(userId: userId)
            if var detail = userDetail {
                userDetail = UserDetailResponse(
                    userId: detail.userId,
                    username: detail.username,
                    isAdmin: detail.isAdmin,
                    isBanned: true,
                    submissionCount: detail.submissionCount,
                    firstSeenAt: detail.firstSeenAt,
                    lastSeenAt: detail.lastSeenAt,
                    ownedBikeCount: detail.ownedBikeCount,
                    leaderboardRanks: detail.leaderboardRanks
                )
            }
        } catch {
            self.error = error.localizedDescription
        }
        isBanning = false
    }

    public func unbanUser() async {
        isBanning = true
        do {
            try await apiClient.unbanUser(userId: userId)
            if var detail = userDetail {
                userDetail = UserDetailResponse(
                    userId: detail.userId,
                    username: detail.username,
                    isAdmin: detail.isAdmin,
                    isBanned: false,
                    submissionCount: detail.submissionCount,
                    firstSeenAt: detail.firstSeenAt,
                    lastSeenAt: detail.lastSeenAt,
                    ownedBikeCount: detail.ownedBikeCount,
                    leaderboardRanks: detail.leaderboardRanks
                )
            }
        } catch {
            self.error = error.localizedDescription
        }
        isBanning = false
    }

    public func clearError() {
        error = nil
    }
}
